compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -1508,15 +1511,39 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
         let lang_item = match (e1, e2, lims) {
             (None, None, HalfOpen) => hir::LangItem::RangeFull,
-            (Some(..), None, HalfOpen) => hir::LangItem::RangeFrom,
+            (Some(..), None, HalfOpen) => {
+                if NEW_RANGE {
+                    hir::LangItem::RangeFromCopy
+                } else {
+                    hir::LangItem::RangeFrom
+                }
+            }
             (None, Some(..), HalfOpen) => hir::LangItem::RangeTo,
-            (Some(..), Some(..), HalfOpen) => hir::LangItem::Range,
+            (Some(..), Some(..), HalfOpen) => {
+                if NEW_RANGE {
+                    hir::LangItem::RangeCopy
+                } else {
+                    hir::LangItem::Range
+                }
+            }
             (None, Some(..), Closed) => hir::LangItem::RangeToInclusive,
-            (Some(..), Some(..), Closed) => unreachable!(),
+            (Some(..), Some(..), Closed) => {
+                if NEW_RANGE {
+                    hir::LangItem::RangeInclusiveCopy
+                } else {
+                    unreachable!() // Handled by lower_expr_range_closed
```

Comment:
nit: could you call `lower_expr_range_closed` here instead of in the match above?

Reply:
Yeah that's much cleaner, good point.